Documentation page for gRPC (#1218)
square/wire: docs/wire_grpc.md

Wire gRPC
=============================

The Wire compiler will generate interfaces for your service RPCs defined in your protobuf schema;
under the condition that the target is set to Kotlin.

Getting Started
---------------

> You can configure how your services are generated via our [Gradle plugin][gradlePlugin]:

Reviewer: ```suggestion
You can configure how your services are generated via the [Gradle plugin][gradlePlugin]:
```

```groovy
wire {
  kotlin { // Kotlin target will generate code for services.
    // `client` to generate interfaces best suited to sending outbound calls.
    // `server` to generate interfaces best suited to receiving inbound calls.
    rpcRole = 'server'
    // Server only
    // `suspending` to generate coroutines APIs that require a Kotlin coroutines context.
    // `blocking` to generate blocking APIs callable by Java and Kotlin.
    rpcCallStyle = 'suspending'
    // Server only

>     // True for emitted services to implement one interface per RPC. 

Reviewer: ```suggestion
    // True for emitted services to generate one interface per RPC. 
```

    singleMethodServices = false
  }
}
```

The generated code varies depending on your RPC role, namely client or server.  

> We'll use the following schema to understand how they differ:

Reviewer: ```suggestion
We'll use the following schema to demonstrate how they differ:
```

```groovy
wire {
  kotlin {
    rpcRole = 'server'
    rpcCallStyle = 'blocking'
  }
}
```
the Wire compiler will generate the following interface for the server to implement:
```kotlin
interface RouteGuideBlockingServer : Service {
  fun GetFeature(request: Point): Feature

  fun ListFeatures(request: Rectangle, response: MessageSink<Feature>)

  fun RecordRoute(request: MessageSource<Point>): RouteSummary

  fun RouteChat(request: MessageSource<RouteNote>, response: MessageSink<RouteNote>)
}
```
For streaming connections, the server can use Wire's blocking APIs: [MessageSource][messageSource]
and [MessageSink][messageSink].

### Suspending APIs

Following this configuration:
```groovy
wire {
  kotlin {
    rpcRole = 'server'
    rpcCallStyle = 'suspending'
  }
}
```
the Wire compiler will generate the following interface for the server to implement:
```kotlin
interface RouteGuideServer : Service {
  suspend fun GetFeature(request: Point): Feature

  fun ListFeatures(request: Rectangle, response: SendChannel<Feature>)

  fun RecordRoute(request: ReceiveChannel<Point>): RouteSummary

  fun RouteChat(request: ReceiveChannel<RouteNote>, response: SendChannel<RouteNote>)
}
```

> The server can use Kotlin coroutines suspend and channels mechanism to execute suspending network

Reviewer: ```suggestion
The server can use Kotlin coroutines' suspend and Channels mechanisms to execute suspending network
```

calls.

### Sample

> Check out our [sample project][sampleProject] demonstrating duplex communications between one

Reviewer: ```suggestion
Check out our [sample project][sampleProject] demonstrating bidirectional communication between an
```

> Android and one server application.

Reviewer: ```suggestion
Android application and a server.
```

### Get it

Use Wire runtime Gradle coordinates:
```groovy
implementation("com.squareup.wire:wire-runtime:LATEST_VERSION")
```

In addition, to use Wire gRPC for clients, use these new Gradle coordinates:

```groovy
implementation("com.squareup.wire:wire-grpc-client:LATEST_VERSION")
```    

>  [gradlePlugin]: https://square.github.io/wire/wire_compiler/

Author: Can't I just shortcut all those `https://square.github.io/wire/` with relative path?

 [grpcCall]: https://square.github.io/wire/3.x/wire-grpc-client/com.squareup.wire/-grpc-call/
 [grpcClient]: https://square.github.io/wire/3.x/wire-grpc-client/com.squareup.wire/-grpc-client/
 [grpcStreamingCall]: https://square.github.io/wire/3.x/wire-grpc-client/com.squareup.wire/-grpc-streaming-call/
 [messageSink]: https://square.github.io/wire/3.x/wire-runtime/com.squareup.wire/-message-sink/
 [messageSource]: https://square.github.io/wire/3.x/wire-runtime/com.squareup.wire/-message-source/

>  [sampleProject]: https://github.com/square/wire/tree/master/wire-grpc-sample

Author: Can't I just shortcut all those https://square.github.io/wire/ with relative path?
